src/libcore/task/wake.rs

```diff
@@ -131,9 +129,14 @@ impl RawWakerVTable {
     /// resources that are associated with this instance of a [`RawWaker`] and
     /// associated task.
     #[rustc_promotable]
-    #[unstable(feature = "futures_api",
-            reason = "futures in libcore are unstable",
-            issue = "50547")]
+    #[cfg_attr(stage0, unstable(feature = "futures_api_const_fn_ptr", issue = "50547"))]
+    #[cfg_attr(not(stage0), stable(feature = "futures_api", since = "1.35.0"))]
+    // `rustc_allow_const_fn_ptr` is a hack that should not be used anywhere else
+    // without first consulting with T-Lang.
+    //
+    // FIXME: remove whenever we have a stable way to accept fn pointers from
+    // const fn (see rfcs/pull/2632)
+    #[cfg_attr(not(stage0), rustc_allow_const_fn_ptr)]
     pub const fn new(
         clone: unsafe fn(*const ()) -> RawWaker,
```

Comment:
I'm a bit surprised we're stabilizing this implementation detail, I would've expected an actual trait to be involved, and these `fn` pointers to be obtained internally (which I think is allowed? oh but `const fn` can't have trait bounds, hmpf).

Also, `*const ()` in a public API? Seems a bit off, I thought we were using `extern type` pointers nowadays?

<hr/>

Anyway, regarding the matter at hand, which is constructors that happen to have banned types, I think this is how we can allow them without allowing the same types in any non-constructor `const fn`: https://github.com/rust-rfcs/const-eval/issues/19#issuecomment-472799062 (by making them pattern aliases)

It also means calls would be promotable, which is good, right?

Reply:
Actually, there is no reason for this to this unsafe.

While the fields of this struct *have* to be `unsafe fn(*const ())`, the arguments could be `fn(&mut T)` where the constructor is generic over `T`.
Sadly, this can't be directly coupled with `RawWaker`'s constructor, but that's only because `RawWakerVTable` isn't itself generic... `RawWaker` could store a `RawWakerVTable<SomePrivateExternType>`, and it would all be much safer.

I'm sorry I haven't remarked this before, I kept thinking this "raw vtable" stuff was kept internal so therefore no need to make it more contrived for safety, I guess I was wrong...

Reply:
Bonus: you could pass a closure where a `fn(&mut T)` is expected, it will just work!

Reply:
I think making the arguments `&mut` would make [this implementation](https://github.com/Nemo157/embrio-rs/blob/master/embrio-executor/src/waker/default.rs) unsound since it uses the same data pointer for all instances.

Reply:
This was already discussed on the RFC (why it isn't a trait and why futures-rs will provide a safe trait for it).

Reply:
@cramertj oops, I should've made separate comments, your resolving meant people wouldn't see my approval of `rustc_allow_const_fn_ptr` (https://github.com/rust-lang/rust/pull/59739#discussion_r275743787)